### ISSUE TITLE
Memory error and large memory usage for parallelised `tl.recover_dynamics`

### ISSUE THREAD

**Issue by WeilerP** [theislab/scvelo#405]
## Description

Running `tl.recover_dynamics` in parallel on a large dataset can lead to memory errors or large RAM usage. See e.g. [here](https://github.com/theislab/scvelo/issues/247#issuecomment-803006512) and [here](https://github.com/theislab/scvelo/issues/247#issuecomment-803020366).
This is likely due to the fact that the the `DynamcisRecovery` class uses the entire AnnData object as an attribute instead of only the minimally required data.

**Comment by WeilerP** [comment]
> Hi,
> I. have a similar problem, I am running a matrix after filtering, which is big. ` n_obs × n_vars = 96344 × 36601`
> I am running this on local on Mac (12 cores, 64GB of RAM), got into memory error.
> I am now running on the cluster and can ask more core, What is the best way to ask for more resources
> 
> `scv.tl.recover_dynamics(adata, n_jobs=32)`

@preetida, you could try reducing `adata` to include only the minimally required information for recovering dynamics. If I am not mistaken, this should only be the imputed data, connectivities and potentially the fitted parameters specified in `adata.var`. You also only need to keep the (potential) velocity genes, for example estimated by the steady state model. This may reduce the data size quite a bit and should reduce the runtime.

**Comment by stela2502** [comment]
@WeilerP reducing the adata to the minimal required seams a good plan, but why is that not handled internally. I do not know what is important for this analysis and to be honest - I do not want to know. In addition that might change in the future and hence the subsetting should definitely be handled internally - right?
I assume that is how the improvement would be implemented?

**Comment by WeilerP** [comment]
> @WeilerP reducing the adata to the minimal required seams a good plan, but why is that not handled internally. I do not know what is important for this analysis and to be honest - I do not want to know. In addition that might change in the future and hence the subsetting should definitely be handled internally - right? I assume that is how the improvement would be implemented?

No, this is not how the improvement is implemented. The improvement will not modify the passed AnnData but only grab the needed data and pass it to the different processes. We do not know which data needs and which not, *i.e.*, we cannot simply remove everything not needed for velocity estimation.

**Comment by stela2502** [comment]
Quite interesting top output after reducing the adata object to the variable genes only.
Now the whole process does work and not fail during the  tl.veclocity run, but every new call to a scv function seams to add one more python full blown process:
```
   PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                
## Big                                                                                                                                                                                
296775 stefanl   20   0  206.4g 151.5g   3376 S   0.0  40.3   0:00.39 python3                                                                                                                                                                                                
296321 stefanl   20   0  551.4g 147.4g  29548 S  49.7  39.2  21:38.44 python3                                                                                                                                                                                                
297532 stefanl   20   0  471.7g 147.3g   3704 S  72.5  39.2   0:05.87 python3                                                                                                                                                                                                
297520 stefanl   20   0  214.4g 146.5g   3656 S   0.0  38.9   0:00.47 python3                                                                                                                                                                                                
## Small
297541 stefanl   20   0   50.3g 777308 363608 R  84.4   0.2   0:17.07 python3                                                                                                                                                                                                
297542 stefanl   20   0   50.3g 746320 367944 R  80.1   0.2   0:16.17 python3                                                                                                                                                                                                
297551 stefanl   20   0   50.3g 744948 362232 R  80.1   0.2   0:14.44 python3                                                                                                                                                                                                
297548 stefanl   20   0   50.3g 739148 357872 R  78.1   0.2   0:15.02 python3                                                                                                                                                                                                
297557 stefanl   20   0   50.2g 727600 359740 R  86.1   0.2   0:13.01 python3                                                                                                                                                                                                
297545 stefanl   20   0   50.2g 727124 360136 R  78.1   0.2   0:15.63 python3                                                                                                                                                                                                
297560 stefanl   20   0   50.2g 712260 357840 R  85.4   0.2   0:12.18 python3                                                                                                                                                                                                
297554 stefanl   20   0   50.2g 711524 359228 R  79.8   0.2   0:13.42 python3                                                                                                                                                                                                
297563 stefanl   20   0   50.2g 701512 359584 R  78.1   0.2   0:11.25 python3                                                                                                                                                                                                
297566 stefanl   20   0   50.2g 684812 343448 R  78.1   0.2   0:10.56 python3                                                                                                                                                                                                
297569 stefanl   20   0   50.2g 684588 354720 R  78.1   0.2   0:09.82 python3                                                                                                                                                                                                
297572 stefanl   20   0   50.2g 679212 358108 R  77.5   0.2   0:08.96 python3                                                                                                                                                                                                
297575 stefanl   20   0   50.2g 631260 303916 R  83.8   0.2   0:09.32 python3                                                                                                                                                                                                
297584 stefanl   20   0   50.2g 627172 317204 R  97.4   0.2   0:06.33 python3                                                                                                                                                                                                
297578 stefanl   20   0   50.2g 617436 303696 R  77.8   0.2   0:07.42 python3                                                                                                                                                                                                
297581 stefanl   20   0   50.2g 603036 292576 R  86.8   0.2   0:06.95 python3                                                                                                                                                                                                
297587 stefanl   20   0   50.2g 540360 239504 R  78.5   0.1   0:04.91 python3                                                                                                                                                                                                
297590 stefanl   20   0   50.2g 371392  77224 R  83.8   0.1   0:04.00 python3                                                                                                                                                                                                
297593 stefanl   20   0   33.9g 264404  47560 R  84.4   0.1   0:03.12 python3                                                                                                                                                                                                
297596 stefanl   20   0   48.8g 220612  45292 R  66.9   0.1   0:02.57 python3                                                                                                                                                                                                                                                   
296782 stefanl   20   0 1390532  24820   6612 S   0.0   0.0   0:05.20 python3 
```
I think it extremely interesting that this easily reaches over 100% memory usage.
Seams to be a linked and not copied memory. So possibly not that much of a problem after all.
